src/hw/actions/initSell.js

```diff
@@ -0,0 +1,246 @@
+      };
+    case "init-swap-requested":
+      return { ...state, initSwapRequested: true, isLoading: false };
+    case "init-swap-result":
+      return {
+        ...state,
+        initSwapResult: e.initSwapResult,
+        isLoading: false,
+      };
+  }
+  return state;
+};
+
+function useFrozenValue<T>(value: T, frozen: boolean): T {
+  const [state, setState] = useState(value);
+  useEffect(() => {
+    if (!frozen) {
+      setState(value);
+    }
+  }, [value, frozen]);
+  return state;
+}
+
+export const createAction = (
+  connectAppExec: (ConnectAppInput) => Observable<ConnectAppEvent>,
+  getTransactionId: (InitSwapInput) => Observable<SwapRequestEvent>,
+  checkSignatureAndPrepare: (CSAPInput) => Observable<CSAPEvent>,
+  onTransactionId: (string) => void
+): InitSwapAction => {
+  const useHook = (
+    reduxDevice: ?Device,
+    initSwapRequest: InitSwapRequest
+  ): InitSwapState => {
+    const [state, setState] = useState(initialState);
+    const [coinifyContext, setCoinifyContext] = useState(null);
+
+    const reduxDeviceFrozen = useFrozenValue(
+      reduxDevice,
+      state.freezeReduxDevice
+    );
+
+    const appState = createAppAction(connectAppExec).useHook(
+      reduxDeviceFrozen,
+      {
+        appName: "Exchange",
+      }
+    );
+
+    const { device, opened } = appState;
+    const { transaction, parentAccount, account } = initSwapRequest;
+
+    useEffect(() => {
+      if (!opened || !device) {
+        setState(initialState);
+        return;
+      }
+
+      const sub = concat(
+        of({ type: "init-swap" }),
+        getTransactionId({
+          deviceId: device.deviceId,
+        }).pipe(
+          map((txId) => ({
+            type: "get-transaction-id",
+            value: txId,
+          }))
+        )
+      )
+        .pipe(
+          tap((e) => {
+            if (e && e.type === "get-transaction-id") {
+              console.log("GET T ID CALLED", onTransactionId);
+              onTransactionId(e.value).then((context) => {
+                const bridge = getAccountBridge(account, parentAccount);
+                const t = bridge.createTransaction(account);
+                const mainAccount = getMainAccount(account, parentAccount);
+                const currency = getAccountCurrency(mainAccount);
+
+                const transaction = bridge.updateTransaction(t, {
+                  amount: parseCurrencyUnit(
+                    currency.units[0],
+                    context.inAmount.toString()
+                  ),
+                  recipient: context.transferIn.details.account,
+                });
+
+                bridge.prepareTransaction(mainAccount, transaction).then(() => {
+                  bridge
+                    .getTransactionStatus(mainAccount, transaction)
+                    .then((status) => {
+                      setCoinifyContext({ context, transaction, status });
+                    });
+                });
+              });
```

Comment:
The promise that goes out on a tap asynchronous, if there is an error would never get seen by the user land, so this would need to be turned to a mergeMap or similar